### DISCUSSION TITLE
A better approach to "await using" and IAsyncDisposable with ConfigureAwait

### DISCUSSION THREAD

**Discussion by RichardR-cg** [dotnet/csharplang#8778] [Language Ideas]
Surely there has to be a better way of making "await using", IAsyncDisposable and ConfigureAwait play nicely. Maybe I've missed a simple solution to this, but if I have, it certainly isn't obvious, and I'd welcome the enlightenment.

Failing that, I'm hoping that the clever people on the C# language team can provide some nicer "syntactic sugar" to wrap that up more neatly in a future version of C#. I don't have a proposed solution, but the situation is described below.

---

I love the IDisposable interface and the "using" block. It limits the scope of things, and ensures that resources being used are disposed.

I also like that if you're using analyzers, it can prompt you to use "using" when you might not otherwise notice that the thing you're using is IDisposable via [CA2000: Dispose objects before losing scope (code analysis) - .NET | Microsoft Learn](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2000).
 
I'm not a huge fan of the change to allow "using" statements instead of blocks, as it relaxes the scope (sometimes with unintended consequences due to the extending of the scope of that variable, and therefore changing the point at which it is disposed); which becomes relevant later.

Along came `async`, and in C#8, `IAsyncDisposable` was introduced ([Implement a DisposeAsync method | Microsoft Learn](https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync)), and although the second paragraph starts by saying...
> It is typical when implementing the IAsyncDisposable interface that classes will also implement the IDisposable interface.

...that certainly isn't always the case (I was recently using a Microsoft class which implements IAsyncDisposable but not IDisposable).

IAsyncDispoable means that we can do `await using`... except that the correct way of writing that (shown in the link above) isn't what you'd expect. The IAsyncDisposable article linked above gives an example of how to write it:

``` csharp
        var exampleAsyncDisposable = new ExampleAsyncDisposable();
        await using (exampleAsyncDisposable.ConfigureAwait(false))
        {
            // Interact with the exampleAsyncDisposable instance.
        }
```

There is a drive to use ConfigureAwait (for reasons given in [CA2007: Do not directly await a Task (code analysis) - .NET | Microsoft Learn](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2007)), which changes the return type.

Notice that this means that the variable which is IAsyncDisposable is no longer constrained by the scope of the using block. So you _could_ write code which uses that variable after the using block, the code probably wouldn't work; but it would compile. The solution to that would be that you have to wrap the above code in braces (just braces, to group the lines shown above together, no special keywords required). Yeuch, and an extra level of indent, and probably some analyzers telling you to remove the "redundant" braces.
 
Interestingly, unlike CA2000 for IDisposable the analyzers do not currently point out if you have something which is IAsyncDisposable, and if you aren't doing an "await using" on it (or its ConfigureAwait), so you need to keep an eye out for this yourself. Perhaps it was the added complexity of the ConfigureAwait syntax which made it hard for them to get the Analyzer to detect (I'm sure it's possible, but it won't be trivial).  The 4 examples given in https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync#stacked-usings (3 acceptable, and one unacceptable) certainly show that it would be tricky to write that analyzer properly.

So it almost seems like the best way to write this is:

``` csharp
        { // These braces serve no other purpose than scoping the variable on the next line.
            var exampleAsyncDisposable = new ExampleAsyncDisposable();
            await using exampleAsyncDisposable.ConfigureAwait(false);
            // Interact with the exampleAsyncDisposable instance.
        }
```

...since this at least doesn't require an extra level of indentation; it's just weird that the thing you're doing the using for is just part of the braced code, rather than it being explicitly clear that that's why the braces are there (like you'd have with a "using block"). 
 
And finally... to cap it all, it would appear that if you use a class which implements both IDisposable and IAsyncDisposable, then it isn't possible to write code which isn't complained about by either CA2000 or CA2007: [CA2000 emitted when await using (false positive) · Issue #3042 · dotnet/roslyn-analyzers (github.com)](https://github.com/dotnet/roslyn-analyzers/issues/3042). I don't have a solution for that, I'm just suppressing the one that I think is least bad to ignore (CA2007).

**Comment by colejohnson66** [comment]
Related: #2661

**Comment by NVedsted** [comment]
At the very least, while waiting for a better solution, perhaps `ConfiguredAsyncDisposable` could provide a way to access the wrapped disposable. So it could work something like this:
```cs
await using (var configuredDisposable = new ExampleAsyncDisposable().ConfigureAwait(false))
{
    var exampleAsyncDisposable = configuredDisposable.Source;
    // Interact with the exampleAsyncDisposable instance.
}
```

We have considered introducing something like this ourselves, but I wonder if there are any drawbacks we might have to consider?

**Reply by colejohnson66** [reply]
That would be a runtime request, which goes in [@dotnet/runtime](https://github.com/dotnet/runtime).

**Comment by NetherGranite** [comment]
Since the problem is that the variable we want to use and the value that should be `await using`'d are different, while it wouldn't be pretty, one possible solution is to just borrow from common `for` loop syntax:

```cs
await using (var theAsyncDisposableWeWant = GetAsyncDisposable(); theAsyncDisposableWeWant.ConfigureAwait(false))
{
    ...
}
```

This doesn't work with the using statement though...
